src/bgw/job_stat.c

```diff
@@ -0,0 +1,345 @@
+
+	if (DatumGetInt32(DirectFunctionCall2(interval_cmp, ival, ival_max)) > 0)
+		ival = ival_max;
+
+	return DatumGetTimestampTz(DirectFunctionCall2(timestamptz_pl_interval, TimestampTzGetDatum(last_finish), ival));
+}
+
+/* For crashes, the logic is the similar as for failures except we also have
+*  a minimum wait after a crash that we wait, so that if an operator needs to disable the job,
+*  there will be enough time before another crash.
+*/
+static
+TimestampTz
+calculate_next_start_crash(int consecutive_crashes, BgwJob *job)
+{
+	TimestampTz now = GetCurrentTimestamp();
+	TimestampTz failure_calc = calculate_next_start_failure(now, consecutive_crashes, job);
+	TimestampTz min_time = TimestampTzPlusMilliseconds(now, MIN_WAIT_AFTER_CRASH_MS);
+
+	if (min_time > failure_calc)
+		return min_time;
+	return failure_calc;
+}
+static bool
+bgw_job_stat_tuple_mark_end(TupleInfo *ti, void *const data)
+{
+	JobResult  *res = data;
+	HeapTuple	tuple = heap_copytuple(ti->tuple);
+	FormData_bgw_job_stat *fd = (FormData_bgw_job_stat *) GETSTRUCT(tuple);
+	Interval   *duration;
+
+	fd->last_finish = GetCurrentTimestamp();
+
+
+
+	duration = DatumGetIntervalP(DirectFunctionCall2(timestamp_mi, TimestampTzGetDatum(fd->last_finish), TimestampTzGetDatum(fd->last_start)));
+	fd->total_duration = *DatumGetIntervalP(DirectFunctionCall2(interval_pl, IntervalPGetDatum(&fd->total_duration), IntervalPGetDatum(duration)));
+
+
+	/* undo marking created by start marks */
+	fd->last_run_success = res->success;
+	fd->total_crashes--;
+	fd->consecutive_crashes = 0;
+
+	if (res->success)
+	{
+		fd->total_success++;
+		fd->consecutive_failures = 0;
+		/* Mark the next start at the end if the job itself hasn't */
+		if (!bgw_job_stat_next_start_was_set(fd))
+			fd->next_start = calculate_next_start_success(fd->last_finish, res->job);
+	}
+	else
+	{
+		fd->total_failures++;
+		fd->consecutive_failures++;
+
+		/*
+		 * Mark the next start at the end if the job itself hasn't (this may
+		 * have happened before failure)
+		 */
+		if (!bgw_job_stat_next_start_was_set(fd))
+			fd->next_start = calculate_next_start_failure(fd->last_finish, fd->consecutive_failures, res->job);
+	}
+
+	catalog_update(ti->scanrel, tuple);
+	heap_freetuple(tuple);
+	return false;
+}
+
+static bool
+bgw_job_stat_tuple_set_next_start(TupleInfo *ti, void *const data)
+{
+	TimestampTz *next_start = data;
+	HeapTuple	tuple = heap_copytuple(ti->tuple);
+	FormData_bgw_job_stat *fd = (FormData_bgw_job_stat *) GETSTRUCT(tuple);
+
+	fd->next_start = *next_start;
+
+	catalog_update(ti->scanrel, tuple);
+	heap_freetuple(tuple);
+	return false;
+}
+
+static bool
+bgw_job_stat_insert_mark_start_relation(Relation rel,
+										int32 bgw_job_id)
+{
+	TupleDesc	desc = RelationGetDescr(rel);
+	Datum		values[Natts_bgw_job_stat];
+	bool		nulls[Natts_bgw_job_stat] = {false};
+	CatalogSecurityContext sec_ctx;
+
+	Interval	zero_ival = {.time = 0,};
+
+	values[Anum_bgw_job_stat_job_id - 1] = Int32GetDatum(bgw_job_id);
```

Comment:
It would be nice to have a `AnumToIndex` macro so for self-documenting-code purposes.